### PR TITLE
New Eclipse SmartHome event mechanism.

### PR DESCRIPTION
**Pull request by sbussweiler**
Breaks JAVA and SSE API.

Also-by: Dennis Nobel <d.nobel@external.telekom.de>
Signed-off-by: Stefan Bussweiler <s.bussweiler@external.telekom.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java [hunk 1]
```diff
@@ -140,23 +144,19 @@ public void statusUpdated(Thing thing, ThingStatusInfo thingStatus) {
             if(thing.getStatusInfo().equals(thingStatus)) {
                 return;
             }
-            
-            thing.setStatusInfo(thingStatus);
+            setThingStatus(thing, thingStatus);
             logger.debug("Status of {} changed to {}", thing.getUID(), thingStatus.toString());
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java]: This log message is redundant due to the EventLogger and should thus be removed.

File: bundles/io/org.eclipse.smarthome.io.monitor/src/main/java/org/eclipse/smarthome/io/monitor/internal/EventLogger.java [hunk 2]
```diff
@@ -7,24 +7,50 @@
...
-        logger.info("{} state updated to {}", itemName, newStatus);
+    public void receive(Event event) {
+        Logger logger = getLogger(event.getType());
+        logger.trace("Received event of type '{}' under the topic '{}' with payload: '{}'", event.getType(),
+                event.getTopic(), event.getPayload());
+        logger.info(event.toString());
+    }
+
+    private Logger getLogger(String eventType) {
+        String loggerName = "runtime.busevents." + eventType;
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.monitor/src/main/java/org/eclipse/smarthome/io/monitor/internal/EventLogger.java]: Not directly related to this PR, but may I ask you to change this to "smarthome." + eventType? Makes the log better readable. Remember to also adapt the logback_debug.xml.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/events/ThingStatusInfoEvent.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+     * 
+     * @return the thing status info
+     */
+    public ThingStatusInfo getStatusInfo() {
+        return thingStatusInfo;
+    }
+
+    @Override
+    public String toString() {
+        return "Status of thing '" + thingUID + "' updated to " + thingStatusInfo.getStatus() + " with detail "
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/events/ThingStatusInfoEvent.java]: You are not even logging the StatusInfo description here, although its main purpose was for logging :-)
I'd suggest to simply change this to:
```
return thingUID + "' updated: " + thingStatusInfo.toString();
```

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemStateEvent.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+     * 
+     * @return the item state
+     */
+    public State getItemState() {
+        return itemState;
+    }
+
+    @Override
+    public String toString() {
+        return "State of item '" + itemName + "' updated to " + itemState;
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemStateEvent.java]: As for the thing status, I would suggest to change this to:
```
return itemName + " updated to " + itemState.toString();
```
Note that the information that it is about a status update is already contained in the logger/class name itself.

**Comment by dnobel** [comment]
Merged directly as contributor is working in the same company as committer.

**Comment by kaikreuzer** [comment]
@sbussweiler Seems that a few references to core.library are not addressed within this PR. Could you please check and fix it with a follow-up PR?

**Comment by kaikreuzer** [comment]
FTR: https://www.eclipse.org/forums/index.php?t=msg&th=1068116&goto=1700619&

**Comment by kaikreuzer** [comment]
@maggu2810, @cdjackson: I would plan to release a new stable ESH version this afternoon, which means that this change becomes available in openHAB2. I have already prepared all adaptions there and will merge them immediately after the ESH update.
As this PR breaks the existing event API and thus might cause small changes on your code, I want to check back with you if you have any veto?

**Comment by maggu2810** [comment]
I'm on vacation, so I can't do further checks.
But I already comment the old PR https://github.com/eclipse/smarthome/pull/355#issuecomment-115237089 and it seems at this time okay for me.

**Pull request merged**